### PR TITLE
Null check baseDirectory before adding to keyFileSearchPaths

### LINKED ISSUE #30349: Passing null into (CSharp|VisualBasic)CommandLineParser.Parse as a base directory gives invalid key paths
If you pass in null as a base directory to our CommandLineParsers.Parse, they add that null into the KeyFileSearchPaths (which is invalid). It's unclear to me if that's a bug (and it shouldn't add them), or I shouldn't be calling with null in the first place.

### PR DESCRIPTION
**Pull request by RikkiGibson**
Resolves #30349 

Seems like the decision is that usages of the baseDirectory parameter should be null-resilient. This null checks the baseDirectory before adding to keyFileSearchPaths so that down the line it doesn't attempt to search a `null` for the key file.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 1]
```diff
@@ -1225,7 +1225,11 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
 
             // Dev11 searches for the key file in the current directory and assembly output directory.
             // We always look to base directory and then examine the search paths.
-            keyFileSearchPaths.Add(baseDirectory);
+            if (!string.IsNullOrWhiteSpace(baseDirectory))
+            {
+                keyFileSearchPaths.Add(baseDirectory);
+            }
+
             if (baseDirectory != outputDirectory)
```
Comment (jasonmalinowski) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Note that if baseDirectory is non-null, but outputDirectory _is_ null, we'll end up with a null path in there again.

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 2]
```diff
@@ -243,6 +243,13 @@ public void ResponseFiles_RelativePaths()
             Assert.Equal(basePath, args.BaseDirectory);
         }
 
+        [Fact]
+        public void NullBaseDirectoryNotAddedToKeyFileSearchPaths()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: >NullBaseDirectoryNotAddedToKeyFileSearchPaths [](start = 20, length = 45)

Please look at and test the corresponding VB API as well.

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: It's difficult to construct this situation for testing because including a null in the `args` array causes NRE in the FlattenArgs method. Should I just modify like the following?

```cs
 if (!string.IsNullOrWhiteSpace(baseDirectory))
{
    keyFileSearchPaths.Add(baseDirectory);
    if (baseDirectory != outputDirectory)
    {
        keyFileSearchPaths.Add(outputDirectory);
    }
}
```

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Probably will just null check the outputDirectory before adding as we should add it even if the baseDirectory is null :)

**Comment by RikkiGibson** [comment]
@jasonmalinowski @jcouv @jaredpar addressed feedback and added vb test+bugfix

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 3]
```diff
@@ -1225,8 +1225,12 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
 
             // Dev11 searches for the key file in the current directory and assembly output directory.
             // We always look to base directory and then examine the search paths.
-            keyFileSearchPaths.Add(baseDirectory);
-            if (baseDirectory != outputDirectory)
+            if (!string.IsNullOrWhiteSpace(baseDirectory))
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Use `IsNullOrEmpty` here. It is possible to have directory names that consist of only white space characters. Example is `"\u200A"`

File: src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCommandLineParser.vb [hunk 4]
```diff
@@ -1291,8 +1291,11 @@ lVbRuntimePlus:
 
             ' Dev10 searches for the keyfile in the current directory and assembly output directory.
             ' We always look to base directory and then examine the search paths.
-            keyFileSearchPaths.Add(baseDirectory)
-            If baseDirectory <> outputDirectory Then
+            If Not String.IsNullOrWhiteSpace(baseDirectory) Then
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCommandLineParser.vb]: Same

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 5]
```diff
@@ -1225,8 +1225,12 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
...
+            {
+                keyFileSearchPaths.Add(baseDirectory);
+            }
+
+            if (!string.IsNullOrWhiteSpace(outputDirectory) && baseDirectory != outputDirectory)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Need a test for when `baseDirectory` is `null` but a valid `outputDirectory` is specified. Should ensure it ends up in the options.

**Comment by jaredpar** [comment]
Done with review pass (iteration 4)

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs (hunk 3 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Any concerns about the usage of `IsNullOrWhiteSpace` for `additionalReferenceDirectories` from [before this change](https://github.com/dotnet/roslyn/blob/b8fa3a23b0bdc5b75d656c1b08462df77bd64328/src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs#L1217)?

```cs
if (!string.IsNullOrWhiteSpace(additionalReferenceDirectories))
```

Comment (jaredpar) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: That code is wrong too then. We should change it as well. Guessing the author wasn't aware of this really odd corner case.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Do you want it to change in this PR?

I'll at least add a test which ensures that whitespace-only directory names are added to the keyFileSearchPaths.

Comment (jaredpar) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Since you're here and it's small lets go ahead and change it too.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Changed the usages, however `Parse` will try to resolve the baseDirectory and therefore getting a useful test that operates on a whitespace baseDirectory is a bit of a challenge (might need to create a temp folder in the cwd instead of with Temp.CreateDirectory())?

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 6]
```diff
@@ -46,6 +46,8 @@ internal sealed override CommandLineArguments CommonParse(IEnumerable<string> ar
         /// <returns>a commandlinearguments object representing the parsed command line.</returns>
         public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string baseDirectory, string sdkDirectory, string additionalReferenceDirectories = null)
         {
+            Debug.Assert(baseDirectory == null || PathUtilities.IsAbsolute(baseDirectory));
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: `PathUtilities.IsAbsolute` returns false for a null

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 7]
```diff
@@ -1214,7 +1216,7 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
             }
 
             // add additional reference paths if specified
-            if (!string.IsNullOrWhiteSpace(additionalReferenceDirectories))
+            if (!string.IsNullOrEmpty(additionalReferenceDirectories))
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Changing this and the check on `keyFileSetting` because whitespace-only filenames are still apparently possible

**Comment by RikkiGibson** [comment]
Would like to test a `null` outputDirectory however it's assumed that all elements in `IEnumerable<string> args` are non-null

**Comment by RikkiGibson** [comment]
PTAL @jaredpar @jcouv @jasonmalinowski

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs (hunk 6 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Does VB need a similar assert?

---
In reply to: [223804356](https://github.com/dotnet/roslyn/pull/30383#discussion_r223804356) [](ancestors = 223804356)

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Yes, thanks. Will add

Comment (jcouv) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: >IsAbsolute [](start = 64, length = 10)

Regarding the assertion of absolute path, where does this guarantee come from?
Since this API is public, the user could pass any path they want. Please add a test for that.

Does the rest of our code have a strong assumption of an absolute path? If so, we should be reporting an error to users. Otherwise, we should remove the assertion.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Parse() calls into FlattenArgs(), which calls into ResolveRelativePath(path, baseDirectory). ResolveRelativePath currently contains an [identical assertion](http://source.roslyn.io/#Microsoft.CodeAnalysis/FileSystem/FileUtilities.cs,99) using PathUtilities.IsAbsolute.

Basically, the CommandLineParser has always assumed the baseDirectory would be an absolute path. This new assert is supposed to make the assumption clear to the reader.

Since it's a public API, maybe just a Debug.Assert is not quite right? Would it be better to have the method always throw on a non-absolute baseDirectory and to have a test that ensures it throws?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Since assertions are skipped in release mode, there is also the question of how the underlying code behaves with a relative path. Maybe it works and some users took a dependency on it.

I'm fine with leaving the assertion as-is. 
If the underlying code blows up on relative paths, then changing what exception we produce doesn't seem a great improvement.

File: src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCommandLineParser.vb [hunk 8]
```diff
@@ -73,6 +73,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         ''' <param name="additionalReferenceDirectories">A string representing additional reference paths.</param>
         ''' <returns>A CommandLineArguments object representing the parsed command line.</returns>
         Public Shadows Function Parse(args As IEnumerable(Of String), baseDirectory As String, sdkDirectory As String, Optional additionalReferenceDirectories As String = Nothing) As VisualBasicCommandLineArguments
+            Debug.Assert(baseDirectory Is Nothing OrElse PathUtilities.IsAbsolute(baseDirectory))
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCommandLineParser.vb]: Looks like this is catching some bad code in `VisualBasicProjectOptionsHelper`. It is passing `""` for `baseDirectory` instead of `Nothing`. That is why so many VB tests are failing now. Really didn't think we'd have this many incorrect usages :frown:

**Comment by RikkiGibson** [comment]
@jaredpar required checks are passing, are we good to squash+merge?

**Comment by jaredpar** [comment]
image

**Comment by RikkiGibson** [comment]
boom
